components/script/dom/htmlinputelement.rs

```diff
@@ -687,6 +704,11 @@ impl HTMLInputElementMethods for HTMLInputElement {
         }
     }
 
+    // https://html.spec.whatwg.org/multipage/#dom-textarea/input-select
+    fn Select(&self) {
+        self.dom_select();
```

Comment:
Thanks for starting to review this.

I can see where you are coming from, but personally I think that to try to put this code in `TextInput` would muddy the concerns of `TextInput`. At the moment it doesn't have any DOM node-related code - it is basically a wrapper around a string, providing some extra state about the selection and so on. This makes it easy to unit test and keeps its implementation focused. To put the selection API code inside `TextInput` would require passing in DOM nodes somehow.

As I say, I agree that there is some indirection here, I guess I just don't think it's that big a deal (easy to find the `dom_select` method using grep, ctags, etc...) It's also just how traits work - when you implement e.g. the `Iterator` trait, you get a whole load of extra methods which aren't defined in the same file, and that's the point of it.

There is potentially another option, though, which is to have a `TextControl` object rather than trait (maybe the name could be improved too). So `HTMLInputElement` and `HTMLTextAreaElement` would each do something like (in their constructor):

``` rust
self.text_control = TextControl::new(&self.upcast::<Node>(), self.textinput)
```

(Actually, we'd probably need to pass in more things than this... but you get the idea.)

And then this method would become:

``` rust
fn Select(&self) {
    self.text_control.dom_select();
}
```

Either way, since I'm not changing the existing architecture in this PR, could we perhaps move this discussion elsewhere, to an issue or something?

Reply:
At the very least, we'll need to make the code look less magical. I don't agree that "the ease of grepping for `dom_select`" is a good reason to not surprise future maintainers of the code, since even if I do find the definition in `text_control.rs`, it's not exactly apparent that both `HTMLInputElement` and `HTMLTextAreaElement` implement the `TextControl` trait, until you search both `htmlinputelement.rs` and `htmltextareaelement.rs` for its implementation.

Encapsulating a `TextInput` inside of a `TextControl` struct sounds fine to me, however. In other words, instead of having HTMLInputElement or HTMLTextAreaElement both contain a `TextInput` field, have a `TextControl` as a member field, inside of which contains a `TextInput`.

Reply:
Respectfully, I'm feeling quite frustrated that you're asking me to change this as a precondition of getting my code merged.

I had thought that the architectural question we're talking about had already been decided in #19272 (where you chose to defer the review to @jdm). I've now submitted this quite substantial PR and the only real feedback I've received after waiting several days is to ask me to make an architectural change which is unrelated to the actual PR, other than that it relates to the same area of the code.

I can understand your perspective, and I place a lot of value on easy to follow code too. I don't think either of us is "right", we just have a difference of opinion about the trade-offs of each choice. I'm willing to revisit the code that you're concerned about, but I'd prefer to do so in a separate PR, since this PR doesn't change the existing architecture. (And I think we should also involve @jdm in any such change since he reviewed #19272.)

Reply:
I'm sorry if my intentions were not clear! I did not mean that you need to make architectural changes in my previous comment. My simple wish is to keep the implementation details of `dom_select` within `htmlinputelement.rs` and `htmltextareaelement.rs`. A workable solution here is to not have a default implementation on the trait, and have it implemented on both `htmlinputelement.rs` and `htmltextareaelement.rs`.

Also, the architectural design is not already decided (refer to @jdm's comment [here](https://github.com/servo/servo/pull/19272#issuecomment-345870872)), but for the purpose of this PR, I concede that this is out of scope.

I really appreciate the work that you've done so far, and I understand it could be frustrating when a code review requires you to make huge architectural changes. In the future, may I invite you to #servo at `irc.mozilla.org` to discuss implementation plans, before any huge changes are made?

Reply:
Another idea would be to just leave a comment here explaining that this is implemented by default by the `TextControl` trait.

Reply:
> A workable solution here is to not have a default implementation on the trait, and have it implemented on both htmlinputelement.rs and htmltextareaelement.rs.

Not sure if I've misunderstood, but if you mean to duplicate the ~110 lines of code which implement `dom_select()` and `set_dom_range_text()` in `TextControl`, this seems like much more of a maintainability problem to me than a bit of indirection.

I'm fine with your other suggestion of adding a comment, though, so I've done that. I'm also willing to revisit the architecture of `TextControl` separately, as I already said.

> In the future, may I invite you to #servo at irc.mozilla.org to discuss implementation plans, before any huge changes are made?

I've been in the IRC room a bit, but I don't really understand what you'd prefer I'd done differently here? Do you mean that you'd prefer I'd discussed the #19272 changes in IRC before submitting a PR?